src/Microsoft.AspNetCore.Hosting/Startup/StartupBase.cs

```diff
@@ -16,4 +16,25 @@ public virtual IServiceProvider ConfigureServices(IServiceCollection services)
             return services.BuildServiceProvider();
         }
     }
+
+    public abstract class StartupBase<TContainer> : IStartup
+    {
+        private readonly IServiceProviderFactory _factory;
+
+        public StartupBase(IServiceProviderFactory factory)
+        {
+            _factory = factory;
+        }
+
+        public abstract void Configure(IApplicationBuilder app);
+
+        public IServiceProvider ConfigureServices(IServiceCollection services)
+        {
+            var builder = _factory.CreateBuilder(services);
+            ConfigureContainer((TContainer)builder);
```

Comment:
What happens if the registered `IServiceProviderFactory`'s builder type doesn't match `TContainer`?